Source/Libraries/openXDA.Nodes/Types/DataPusher/DataPusherNode.cs

```diff
@@ -0,0 +1,137 @@
+//  02/08/2021 - Stephen C. Wills
+//       Generated original version of source code.
+//
+//******************************************************************************************************
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Data;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Web.Http;
+using System.Web.Http.Controllers;
+using GSF.Configuration;
+using GSF.Data;
+using GSF.Data.Model;
+using GSF.IO;
+using GSF.Threading;
+using log4net;
+using openXDA.Configuration;
+using openXDA.DataPusher;
+using openXDA.Model;
+
+namespace openXDA.Nodes.Types.DataPusher
+{
+    public class DataPusherNode : NodeBase
+    {
+        #region [ Members ]
+
+        // Nested Types
+        private class Settings
+        {
+            public Settings(Action<object> configure) =>
+                configure(this);
+
+            [Category]
+            [SettingName(DataPusherSection.CategoryName)]
+            public DataPusherSection DataPusherNodeSettings { get; } = new DataPusherSection();
+
+            [Category]
+            [SettingName(FileWatcherSection.CategoryName)]
+            public FileWatcherSection FileWatcherSettings { get; } = new FileWatcherSection();
+        }
+
+        private class DataPusherWebController : ApiController
+        {
+            private DataPusherNode Node { get; }
+
+            public DataPusherWebController(DataPusherNode node) =>
+                Node = node;
+        }
+
+        #endregion
+
+        #region [ Constructors ]
+
+        public DataPusherNode(Host host, Node definition, NodeType type)
+            : base(host, definition, type)
+        {
+            Action<object> configurator = GetConfigurator();
+            ScheduleAllInstances(configurator);
+        }
+
+        #endregion
+
+        #region [ Properties ]
+
+        private DataPusherEngine DataEngine { get; set; }
+
+        #endregion
+
+        #region [ Methods ]
+
+        public override IHttpController CreateWebController() =>
+            new DataPusherWebController(this);
+
+        protected override void OnReconfigure(Action<object> configurator)
+        {
+            ScheduleAllInstances(configurator);
+        }
+
+        private void ScheduleAllInstances(Action<object> configurator)
+        {
+            Settings settings = new Settings(configurator);
+            DataEngine = new DataPusherEngine(() => new AdoDataConnection(settings.DataPusherNodeSettings.ConnectionString));
+            IEnumerable<RemoteXDAInstance> allInstances;
+            using (AdoDataConnection connection = new AdoDataConnection(settings.DataPusherNodeSettings.ConnectionString))
+            {
+                allInstances = new TableOperations<RemoteXDAInstance>(connection).QueryRecordsWhere("Frequency != '*'");
+            }
+            foreach (RemoteXDAInstance instance in allInstances)
```

Comment:
I suppose this probably works because `QueryRecordsWhere()` uses `RetrieveData()` under the covers, but I think it would be wise to use `ToList()` inside the `using` to ensure that the `IEnumerable` is evaluated before the connection is disposed.

Reply:
Wouldn't the scope of the variable allInstances being outside of the using statement ensure this already? for this resolution, I have moved the foreach into the statement along with the variable declaration, but I'm curious as to how it would lose information leaving the using statement?

Reply:
Many instances of `IEnumerable<T>` are lazy evaluated. For instance, imagine if the implementation looked something like this.

```c#
public IEnumerable<T> QueryRecordsWhere(string filterExpression, params object[] parameters)
{
    string query = BuildSelectQuery(filterExpression);

    using (DataTable table = Connection.RetrieveData(query, parameters))
    {
        foreach (DataRow row in table.Rows)
            yield return LoadRecord(row);
    }
}
```

In this case, the call to `Connection.RetrieveData()` wouldn't be evaluated until the user attempts to enumerate the collection with a `foreach` loop. In this case, you'd receive an error if you disposed the connection before attempting to enumerate the result. In cases like this where the behavior depends on the implementation details of the API, you can use `Enumerable.ToList<T>()` to force the `IEnumerable<T>` to evaluate at the point where you turn it into a list and avoid potential errors.